File: LeetCodePractice/src/dfs/MaxAreaOfIsland.java
package dfs;

/*
Given a non-empty 2D array grid of 0's and 1's, an island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.
Find the maximum area of an island in the given 2D array. (If there is no island, the maximum area is 0.)

Example 1:

[[0,0,1,0,0,0,0,1,0,0,0,0,0],
 [0,0,0,0,0,0,0,1,1,1,0,0,0],
 [0,1,1,0,1,0,0,0,0,0,0,0,0],
 [0,1,0,0,1,1,0,0,1,0,1,0,0],
 [0,1,0,0,1,1,0,0,1,1,1,0,0],
 [0,0,0,0,0,0,0,0,0,0,1,0,0],
 [0,0,0,0,0,0,0,1,1,1,0,0,0],
 [0,0,0,0,0,0,0,1,1,0,0,0,0]]
Given the above grid, return 6. Note the answer is not 11, because the island                   must be connected 4-directionally.
Example 2:

[[0,0,0,0,0,0,0,0]]
Given the above grid, return 0.
Note: The length of each dimension in the given grid does not exceed 50.
*/
public class MaxAreaOfIsland {
	int rows;
	int cols;
	int[] dirs = { 0, 1, 0, -1, 0 };
	public int maxAreaOfIsland(int[][] grid) {
		rows = grid.length;
		cols = grid[0].length;
		int res = 0;
		for (int r = 0; r < rows; r++) {
			for (int c = 0; c < cols; c++) {
				if (grid[r][c] == 1) {
					res = Math.max(res, dfs(grid, r, c));
				}
			}
		}
		return res;
	}

	private int dfs(int[][] grid, int r, int c) {
		int res = 1;
		grid[r][c] = 0;
		for (int i = 0; i < dirs.length - 1; i++) {
			int nb_r = r + dirs[i];
			int nb_c = c + dirs[i + 1];
			if (nb_r >= 0 && nb_r < rows && nb_c >= 0 && nb_c < cols && grid[nb_r][nb_c] == 1) {
				res += dfs(grid, nb_r, nb_c);
			}
		}
		return res;
	}
}

File: LeetCodePython/MaxAreaOfIsland.py
"""
You are given an m x n binary matrix grid. An island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.
The area of an island is the number of cells with a value 1 in the island.
Return the maximum area of an island in grid. If there is no island, return 0.

Example 1:
Input: grid = [[0,0,1,0,0,0,0,1,0,0,0,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,1,1,0,1,0,0,0,0,0,0,0,0],[0,1,0,0,1,1,0,0,1,0,1,0,0],[0,1,0,0,1,1,0,0,1,1,1,0,0],[0,0,0,0,0,0,0,0,0,0,1,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,0,0,0,0,0,0,1,1,0,0,0,0]]
Output: 6
Explanation: The answer is not 11, because the island must be connected 4-directionally.

Example 2:
Input: grid = [[0,0,0,0,0,0,0,0]]
Output: 0

Constraints:
m == grid.length
n == grid[i].length
1 <= m, n <= 50
grid[i][j] is either 0 or 1.
"""
from typing import List


class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        res, rows, cols = 0, len(grid), len(grid[0])
        dirs = [0, 1, 0, -1, 0]

        def dfs(x, y):
            ret = 1
            grid[x][y] = 0
            for i in range(len(dirs) - 1):
                nb_x, nb_y = x + dirs[i], y + dirs[i + 1]
                if 0 <= nb_x < rows and 0 <= nb_y < cols and grid[nb_x][nb_y] == 1:
                    ret += dfs(nb_x, nb_y)
            return ret

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    res = max(res, dfs(r, c))
        return res
